demo/torchdynamo/dynamo_utils.py

```diff
@@ -0,0 +1,193 @@
+#  Copyright 2022, Lefebvre Dalloz Services
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import contextlib
+import gc
+import time
+from dataclasses import dataclass
+from typing import Callable, Dict, List, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+import torch
+import torchdynamo
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+from onnxruntime import GraphOptimizationLevel
+from tensorrt.tensorrt import Runtime
+from torch._C._autograd import ProfilerActivity
+from torchdynamo.eval_frame import OptimizeContext
+from transformers import PreTrainedModel
+
+from transformer_deploy.backends.ort_utils import create_model_for_provider, inference_onnx_binding
+from transformer_deploy.backends.trt_utils import load_engine
+
+
+seq_lengths = [16, 64, 128, 256, 384, 512]
+batch_sizes = [1, 8, 16, 32, 64, 128, 256]
+shapes_to_test: Dict[int, List[int]] = {b_s: seq_lengths for b_s in batch_sizes}
+
+
+@dataclass
+class BenchmarkOutput:
+    latency: float
+    output: Dict[str, torch.Tensor]
+
+
+def get_pytorch_input(size: Tuple[int, int]) -> Dict[str, torch.Tensor]:
+    return {
+        "input_ids": torch.randint(2, 1000, size=size, dtype=torch.int32, device="cuda"),
+        "attention_mask": torch.ones(size=size, dtype=torch.int32, device="cuda"),
+    }
+
+
+def benchmark(
+    fn: Callable[[Dict[str, torch.Tensor]], BenchmarkOutput], shapes: Dict[int, List[int]] = shapes_to_test
+) -> np.ndarray:
+    gc.collect()  # delete all deletable objects so CUDA memory can be freed by empty_cache
+    torch.cuda.empty_cache()
+    timings: List[List[float]] = list()
+    for is_warmup in [True, True, False]:
+        for batch_size, seq_lens in shapes.items():
+            batch_timings: List[float] = list()
+            for seq_len in seq_lens:
+                inputs = get_pytorch_input(size=(batch_size, seq_len))
+                latencies = list()
+                nb_retry = 1 if is_warmup else 5
```

Comment:
IMO it's better to execute the exact same code in warmup and simply get the last result

Reply:
ok